modules/detectron/batch_permutation_op.h

```diff
@@ -32,8 +33,37 @@ class BatchPermutationOp final : public Operator<Context> {
   USE_OPERATOR_CONTEXT_FUNCTIONS;
 
   bool RunOnDevice() override {
-    // No CPU implementation for now
-    CAFFE_NOT_IMPLEMENTED;
+    const auto& X = Input(0);
+    auto& indices = Input(1);
+    auto* Y = Output(0);
+
+    CAFFE_ENFORCE(indices.ndim() == 1, "indices must be 1-d");
+    CAFFE_ENFORCE(
+      X.dim32(0) == indices.dim32(0),
+      "X.dim32(0) must be equal to indices.dim32(0)",
+      "(",
+      X.dim32(0),
+      " vs. ",
+      indices.dim32(0),
+      ")");
+
+    Y->ResizeLike(X);
+
+    const int N = X.dim32(0);
+    const int C = X.dim32(1);
+    const int H = X.dim32(2);
+    const int W = X.dim32(3);
+
+    const float *src = X.template data<float>();
+    float *dst = Y->template mutable_data<float>();
+
+    for (int i = 0; i < N; i++) {
+      int idx = indices.template data<int>()[i];
+
+      memcpy(dst + i * C * H * W, src + idx * C * H * W, sizeof(float) * C * H * W);
```

Comment:
If you only want to implement it on CPU. please move the definition of RunOnDevice to cc file

Reply:
std::memcpy(...)

Reply:
Here maybe we can consider to make it in-place.